### DISCUSSION TITLE
why does walletClient not inherit public client

### DISCUSSION THREAD

**Discussion by wighawag** [wevm/viem#649] [General]
Currently walletClient can call writeContract but not read

This is annoying, especially in the context of contract instances as when you create them you need to specify either a public client (which give access to the read method) or a wallet client (which only give access to write method)

Is there a fundamental reason why there could not be a ReadAndWriteClient ?

**Comment by wighawag** [comment]
I am currently using this so I can have a contract instance with both `read` and `write`

```
const contract = {
  ...getContract({
	...contractInfo,
	walletClient,
  }),
  ...getContract({
	...contractInfo,
	publicClient,
  }),
};
```

**Reply by tmm** [reply]
You can pass both `publicClient` and `walletClient` to `getContract`. For more info on why wallet client does not have public actions, see https://github.com/wagmi-dev/viem/blob/main/site/docs/faq.md#why-doesnt-wallet-client-support-public-actions

**Comment by wighawag** [comment]
Thanks! should have tried that :)

thanks for the link to the doc, I hope wallets keep providing more than `eth_sendTransaction` though, We need to keep the user in control and not force devs to provide their own rpc :(
